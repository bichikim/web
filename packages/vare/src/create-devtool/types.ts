export interface ApiSetting {
  foo?: boolean
}

export interface TimeLineInfo {
  color: any
  label: string
}

export interface CreateDevToolOptions {
  /**
   * @default vare-structure
   */
  id?: string

  /**
   * @default Vare
   */
  label?: string

  timeLines?: Record<string, TimeLineInfo>
}
