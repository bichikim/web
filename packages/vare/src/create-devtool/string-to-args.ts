const compact = (value: any[]) => (value.filter(Boolean))

export const stringToArgs = (text: string) => {
  return compact(text.split(',').map((value) => {
    const _value = value.trim()
    if (/^".+"$/u.test(_value)) {
      return _value.replace(/^"+/u, '').replace(/"+$/u, '')
    }
    const number = Number(_value)
    if (Number.isNaN(number)) {
      return null
    }
    return number
  }))
}
