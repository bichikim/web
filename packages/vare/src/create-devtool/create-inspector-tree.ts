import {useInfo} from 'src/info'
import {CustomInspectorNode} from '@vue/devtools-api'
import {stateName} from 'src/state'
import {mutationName} from 'src/mutate'
import {computeName, computeRefName} from 'src/compute'

const textBackgroundColors = {
  action: 0x73_AB_FE,
  atom: 0xC0_4A_C2,
  [computeName]: 0x42_B9_83,
  [computeRefName]: 0x42_B9_83,
  multi: 0xFF_C6_6D,
  [mutationName]: 0xFF_98_4F,
  [stateName]: 0x42_B9_83,
  tip: 0xF8_F8_F8,
  unknown: 0xFF_00_00,
}

export const createInspectorTree = (targets: Record<string, any>) => {
  const info = useInfo()
  const relationMap = new Map<string, any>()

  const nodes: CustomInspectorNode[] = Object.keys(targets).map((name: string) => {
    const target = targets[name]
    const children: any[] = []
    const targetInfo = info.get(target)

    const kind = targetInfo?.kind ?? 'unknown'
    return {
      id: name,
      label: name,
      tags: [{
        backgroundColor: textBackgroundColors[kind],
        label: kind,
        textColor: 0x00_00_00,
      }],
    }
  })

  return {
    nodes,
  }
}
