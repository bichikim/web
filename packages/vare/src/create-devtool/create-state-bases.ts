import {getGlobalInfo, getState} from 'src/info'
import {StateBase} from '@vue/devtools-api'

export const createStateBases = (targets: Record<string, any>): Record<string, StateBase> => {
  const info = getGlobalInfo()

  if (!info) {
    return {}
  }

  return Object.keys(targets).reduce<Record<string, StateBase>>((result, key: string) => {
    const value = targets[key]
    const state = getState(info, value) ?? value

    result[key] = {
      editable: true,
      key,
      objectType: 'reactive',
      value: state,
    }
    return result
  }, {})
}
