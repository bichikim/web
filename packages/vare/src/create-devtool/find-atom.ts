import {isAtom} from 'src/atom'
export const findAtom = (target: any, deeps: string[] = []) => {
  const atoms: any[] = []

  if (typeof target === 'object' && !Array.isArray(target)) {
    Object.keys(target).forEach((key) => {
      const value = target[key]
      atoms.push(...findAtom(value, [...deeps, key]))
    })
  }

  if (deeps.length === 0) {
    return atoms
  }

  if (isAtom(target)) {
    atoms.push([deeps.join('.'), target])
    return atoms
  }

  return atoms
}
