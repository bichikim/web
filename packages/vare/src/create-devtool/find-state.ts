import {isAtom} from 'src/atom'
import {isState} from 'src/state'
export const findState = (target: any, deeps: string[] = []) => {
  const atoms: any[] = []

  if (typeof target === 'object' && !Array.isArray(target)) {
    Object.keys(target).forEach((key) => {
      const value = target[key]
      atoms.push(...findState(value, [...deeps, key]))
    })
  }

  if (deeps.length === 0) {
    return atoms
  }

  if (isAtom(target) || isState(target)) {
    atoms.push([deeps.join('.'), target])
    return atoms
  }

  return atoms
}
