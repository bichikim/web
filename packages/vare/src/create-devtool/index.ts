import {DevtoolsPluginApi, setupDevtoolsPlugin, TimelineEvent} from '@vue/devtools-api'
import {drop, isSSR} from '@winter-love/utils'
import {watchTrigger} from 'src/create-devtool/watch-trigger'
import {VareInfo} from 'src/info'
import {App} from 'vue-demi'
import {createStateBases} from './create-state-bases'
import {ApiSetting, CreateDevToolOptions} from './types'
import {watchState} from './watch-state'
import {atomName} from 'src/atom'
import {createInspectorTree} from './create-inspector-tree'

export const DEVTOOL_ID = 'com.npmjs.packages.vare'

// eslint-disable-next-line max-lines-per-function
export const createDevTool = (
  app: App,
  targets: Record<string, any>,
  options: CreateDevToolOptions = {},
) => {
  if (isSSR()) {
    return
  }
  const stateBases = createStateBases(targets)

  const {
    id: inspectorId = 'vare-structure',
    label = 'Vare',
    timeLines = {
      [atomName]: {
        color: 0xF08D49,
        label: 'Vare Atom Actions',
      },
    },
  } = options

  let _api: DevtoolsPluginApi<ApiSetting>

  setupDevtoolsPlugin({
    app,
    id: DEVTOOL_ID,
    label,
    packageName: 'vare',
  }, (api) => {
    _api = api

    api.addInspector({
      icon: 'mediation',
      id: inspectorId,
      label: `${label} Structure`,
      stateFilterPlaceholder: 'Search for state',
      treeFilterPlaceholder: `Search for ${label}`,
    })

    Object.keys(timeLines).forEach((id: string) => {
      const info = timeLines[id]
      api.addTimelineLayer({
        color: info.color,
        id,
        label: info.label,
      })
    })

    api.on.getInspectorTree((payload) => {
      if (payload.app !== app || payload.inspectorId !== inspectorId) {
        return
      }

      const {nodes} = createInspectorTree(targets)
      console.log(nodes)
      payload.rootNodes = nodes
    })

    api.on.getInspectorState((payload) => {
      if (payload.app !== app || payload.inspectorId !== inspectorId) {
        return
      }

      const state = stateBases[payload.nodeId]

      if (state) {

        payload.state = state.refresh()
      }
    })

    api.on.editInspectorState((payload) => {
      if (payload.app !== app || payload.inspectorId !== inspectorId) {
        return
      }

      const state = stateBases[payload.nodeId]

      if (state) {
        const path = drop(payload.path)
        const _state = state.base
        const {value} = payload.state
        payload.set(_state.value, path, value)
      }
    })
  })

  const updateTimeline = (layerId: string, event: Omit<TimelineEvent, 'time'>, all?: boolean) => {
    _api?.addTimelineEvent({
      all,
      event: {
        ...event,
        time: Date.now(),
      },
      layerId,
    })
  }

  const updateTrigger = (info?: VareInfo<string>) => {
    if (!info) {
      return
    }

    updateTimeline(info.identifier, {
      data: {
        type: info.identifier,
      },
    })
  }

  const updateTree = () => {
    _api?.sendInspectorTree(inspectorId)
  }

  const updateState = () => {
    _api?.sendInspectorState(inspectorId)
  }

  watchState(targets, updateState)
  watchTrigger(targets, updateTrigger)

  return {
    updateState,
    updateTimeline,
    updateTree,
  }
}
