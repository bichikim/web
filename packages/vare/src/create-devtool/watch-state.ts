import {getGlobalInfo, getState, VareInfo} from 'src/info'
import {watch} from 'vue-demi'

export const watchState = (targets?: Record<string, any>, callback?: (info?: VareInfo<any>) => unknown) => {
  const info = getGlobalInfo()

  if (!info || !targets) {
    return
  }

  Object.keys(targets).forEach((key) => {
    const value = targets[key]
    const state = getState(info, value) ?? value

    if (typeof state !== 'object') {
      return
    }

    watch(state, () => {
      callback?.(info.get(value))
    })
  })
}
