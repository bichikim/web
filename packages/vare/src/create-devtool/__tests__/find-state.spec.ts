import {findState} from 'src/create-devtool/find-state'
import {atom} from 'src/atom'
import {state} from 'src/state'

describe('find-state', () => {
  it('should return atoms', () => {
    const result = findState(atom({
      foo: atom({
        bar: 'bar',
        john: atom({
          name: 'john',
        }),
      }),
    }))

    expect(result).toMatchSnapshot()
  })
  it('should return state', () => {
    const result = findState(state({
      bar: 'bar',
      foo: state({
        deep: 'deep-foo',
      }),
    }))

    expect(result).toMatchSnapshot()
  })
})
