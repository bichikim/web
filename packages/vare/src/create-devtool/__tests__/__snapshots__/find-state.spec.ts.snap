// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-state should return atoms 1`] = `
Array [
  Array [
    "foo.john",
    Object {
      "name": "john",
      Symbol(atom): true,
      Symbol(atom action watch): undefined,
    },
  ],
  Array [
    "foo",
    Object {
      "bar": "bar",
      "john": Object {
        "name": "john",
        Symbol(atom): true,
        Symbol(atom action watch): undefined,
      },
      Symbol(atom): true,
      Symbol(atom action watch): undefined,
    },
  ],
]
`;

exports[`find-state should return state 1`] = `
Array [
  Array [
    "foo",
    Object {
      "deep": "deep-foo",
      Symbol(state): true,
    },
  ],
]
`;
